Extension character controller

HaydenHefford opened the pull request:
Extended character controller with the following:

- can now move camera using right joystick on controller, as this was not included in my previous implementation
- can now crouch and uncrouch
- uncrouch uses a raycast check to ensure that the player can only stand when there is appropriate space

HaydenHefford committed:
Added controller rotation

HaydenHefford committed:
Changed controller rotation

Made the rotation a "get_vector" rather than two separate "get_axis", makes it feel a lot smoother

HaydenHefford committed:
Implemented crouch action

Crouch using CTRL for keyboard, L3 for controller

HaydenHefford committed:
Added raycast for uncrouch check

Raycast checks for collision at normal player height - if collision found, you cannot uncrouch

HaydenHefford committed:
Cleared unused variables

bigmonmulgrew commented on Scenes/Characters/player.tscn:
```diff
@@ -1,15 +1,62 @@
 
 [node name="Camera3D" type="Camera3D" parent="."]
-transform = Transform3D(1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0.677103, 0)
+transform = Transform3D(1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0.448461, 0)
```
Player camera height is just under 0.5m high. Consider changing its heights to better match a person's height.
Also note that I just realised the player placeholder is partially phased through the floor due to it being centred so its not a good indicator of height. Might be worth making a dummy that can be free placed and moved independently.

Height should probably be about 1.5m but this might cause some camera issues with door ways, collider will need updating to match too

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -1,22 +1,33 @@
 extends CharacterBody3D
 
 
```
variables should be grouped according to style guide example

Constants
Exports
Object references
Runtime variables

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -1,22 +1,33 @@
+	$RayCast3D.enabled = true
 	Input.set_mouse_mode(Input.MOUSE_MODE_CAPTURED)
+	$CollisionShape3D.shape.height = player_height
+	print($CollisionShape3D.shape.height)
```
print statements used for debugging should either be removed or have a high amount of verbosity.

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -35,8 +46,29 @@ func handle_movement():
+	rotation_degrees.x -= rotation_input.x * joystick_sensitivity
+
+func crouch(delta: float):
+	if Input.is_action_just_pressed("crouch") && (b_is_crouching == false):
```
consider early exit condition to improve readability. This also requires less checks on frames where we do not crouch.

if not Input.is_action_just_pressed("crouch"):
		return
		
	if (b_is_crouching == false):

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -35,8 +46,29 @@ func handle_movement():
+		$AnimationPlayer.play("crouch")
+		print($CollisionShape3D.shape.height)
+	elif Input.is_action_just_pressed("crouch") && (b_is_crouching == true):
+		$RayCast3D.force_raycast_update()
```
Might be worth considering if we really need the forced update, I don't think our current use case will matter if we are a few pixels out.
This would need testing but we could save some performance if we could do away with it.

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -35,8 +46,29 @@ func handle_movement():
+	rotation_degrees.y -= rotation_input.y * joystick_sensitivity
+	rotation_degrees.x -= rotation_input.x * joystick_sensitivity
+
+func crouch(delta: float):
```
delta is unused, if intentional add an underscore, if not then just remove it.

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -35,8 +46,29 @@ func handle_movement():
+		b_is_crouching = true
+		current_move_speed = crouch_walk_speed
+		$AnimationPlayer.play("crouch")
+		print($CollisionShape3D.shape.height)
```
print statements used for debugging should either be removed or have a high amount of verbosity.

bigmonmulgrew commented on Scripts/Player_Scripts/player.gd:
```diff
@@ -35,8 +46,29 @@ func handle_movement():
+			$AnimationPlayer.play_backwards("crouch")
+			print($CollisionShape3D.shape.height)
+		else:
+			print("Not enough space to stand up!")
```
# Maybe play a sound here and remove the print

bigmonmulgrew replied:
I dont know why the font is massive

bigmonmulgrew reviewed (changes requested):
See inline comments

HaydenHefford committed:
Implemented requested changes

Only change not implemented from request is sound when unable to stand up.

Also, camera height is not actually that low - it is that far above the centre of the player. It should not need bringing up further, although it has been bumped up slightly in this while tweaking the camera.
